### PR TITLE
Test if a path is valid

### LINKED ISSUE #66: Wrong order of links in the primary_path leads to strange behavior
Hi all,

If we create an EVC using primary_path attribute with a valid path (i.e., list of links) but with endpoints inside each link in a non-ordered sequence (even in the same order returned by kytos), Kytos-ng/mef_eline presents a strange behavior: the circuit is marked as active, current path includes a valid path, but the circuit does not work and the flows in the switches are very messed up.

Steps to reproduce:

1. Run Kytos (e.g., by using docker container `amlight/kytos:latest`)
2. Run a simple topology:
```
mn --topo linear,5 --controller=remote,ip=127.0.0.1
```
3. Enable all the switches and links:
```
IP=127.0.0.1; for sw in $(curl -s http://$IP:8181/api/kytos/topology/v3/switches | jq -r '.switches[].id'); do curl -H 'Content-type: application/json' -X POST http://$IP:8181/api/kytos/topology/v3/switches/$sw/enable; curl -H 'Content-type: application/json' -X POST http://$IP:8181/api/kytos/topology/v3/interfaces/switch/$sw/enable; done

IP=127.0.0.1; for l in $(curl -s http://$IP:8181/api/kytos/topology/v3/links | jq -r '.links[].id'); do curl -H 'Content-type: application/json' -X POST http://$IP:8181/api/kytos/topology/v3/links/$l/enable; done
```
4. List the links to be used later:
```
# curl -s http://127.0.0.1:8181/api/kytos/topology/v3/links | jq -r '.links[] | .id + " {\"endpoint_a\": {\"id\": \"" +(.endpoint_a.id|tostring) + "\"}, \"endpoint_b\": {\"id\": \"" + (.endpoint_b.id|tostring) + "\"}}"'
0b814adbd3b05669482ca479280787e55be5c155780f8a780423fa9e67e3a848 {"endpoint_a": {"id": "00:00:00:00:00:00:00:05:2"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:04:3"}}
4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30 {"endpoint_a": {"id": "00:00:00:00:00:00:00:02:3"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:03:2"}}
88cfed769fbb073a8ca849394180f6b229cbd1cbadd8a1970e6c721ffd675153 {"endpoint_a": {"id": "00:00:00:00:00:00:00:04:2"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:03:3"}}
cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260 {"endpoint_a": {"id": "00:00:00:00:00:00:00:02:2"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:01:2"}}
```
5. Using the links above, create and EVC with static primary path:
```
# curl -H 'Content-type: application/json' -X POST http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc -d '{"name": "my evc 2000", "dynamic_backup_path": true, "enabled": true, "uni_a": {"tag": {"value": 2000, "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:01:1"}, "uni_z": {"tag": {"value": 2000, "tag_type": 1}, "interface_id": "00:00:00:00:00:00:00:05:1"}, "primary_path": [{"endpoint_a": {"id": "00:00:00:00:00:00:00:02:2"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:01:2"}}, {"endpoint_a": {"id": "00:00:00:00:00:00:00:02:3"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:03:2"}}, {"endpoint_a": {"id": "00:00:00:00:00:00:00:04:2"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:03:3"}}, {"endpoint_a": {"id": "00:00:00:00:00:00:00:05:2"}, "endpoint_b": {"id": "00:00:00:00:00:00:00:04:3"}}]}'
{"circuit_id":"9784971f2e944a"}
```
6. Check the flows:
```
# for s in s1 s2 s3 s4 s5; do echo $s; ovs-ofctl dump-flows $s; done
s1
 cookie=0xaa9784971f2e944a, duration=6.508s, table=0, n_packets=0, n_bytes=0, in_port=3,dl_vlan=352 actions=mod_vlan_vid:1994,output:"s1-eth2"
 cookie=0xaa9784971f2e944a, duration=6.465s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth1",dl_vlan=2000 actions=strip_vlan,mod_vlan_vid:1994,output:"s1-eth2"
 cookie=0xaa9784971f2e944a, duration=6.463s, table=0, n_packets=0, n_bytes=0, in_port="s1-eth2",dl_vlan=1994 actions=strip_vlan,mod_vlan_vid:2000,output:"s1-eth1"
 cookie=0x0, duration=96.221s, table=0, n_packets=541, n_bytes=22722, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
s2
 cookie=0x0, duration=96.171s, table=0, n_packets=1082, n_bytes=45444, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
s3
 cookie=0xaa9784971f2e944a, duration=6.512s, table=0, n_packets=0, n_bytes=0, in_port="s3-eth2",dl_vlan=352 actions=mod_vlan_vid:2072,output:"s3-eth2"
 cookie=0xaa9784971f2e944a, duration=6.505s, table=0, n_packets=0, n_bytes=0, in_port="s3-eth2",dl_vlan=2072 actions=mod_vlan_vid:352,output:"s3-eth2"
 cookie=0xaa9784971f2e944a, duration=6.491s, table=0, n_packets=0, n_bytes=0, in_port="s3-eth3",dl_vlan=2072 actions=mod_vlan_vid:3860,output:"s3-eth2"
 cookie=0xaa9784971f2e944a, duration=6.488s, table=0, n_packets=0, n_bytes=0, in_port="s3-eth2",dl_vlan=3860 actions=mod_vlan_vid:2072,output:"s3-eth3"
 cookie=0x0, duration=96.141s, table=0, n_packets=1082, n_bytes=45444, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
s4
 cookie=0x0, duration=96.109s, table=0, n_packets=1082, n_bytes=45444, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
s5
 cookie=0xaa9784971f2e944a, duration=6.476s, table=0, n_packets=0, n_bytes=0, in_port="s5-eth1",dl_vlan=2000 actions=strip_vlan,mod_vlan_vid:3860,output:3
 cookie=0xaa9784971f2e944a, duration=6.470s, table=0, n_packets=0, n_bytes=0, in_port=3,dl_vlan=3860 actions=strip_vlan,mod_vlan_vid:2000,output:"s5-eth1"
 cookie=0x0, duration=96.028s, table=0, n_packets=541, n_bytes=22722, priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
```
7. Check the status of the EVC:
```
root@7bad6827e2f2:/# curl http://127.0.0.1:8181/api/kytos/mef_eline/v2/evc/9784971f2e944a | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  7160  100  7160    0     0   776k      0 --:--:-- --:--:-- --:--:--  776k
{
  "active": true,
  "archived": false,
  "backup_links": [],
  "backup_path": [],
  "bandwidth": 0,
  "circuit_scheduler": [],
  "creation_time": "2021-10-01T19:35:36",
  "current_path": [
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:02:2",
        "link": "cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260",
        "lldp": true,
        "mac": "7e:74:84:8c:62:f5",
        "metadata": {},
        "name": "s2-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:02",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:01:2",
        "link": "cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260",
        "lldp": true,
        "mac": "4a:57:a9:ed:97:6e",
        "metadata": {},
        "name": "s1-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:01",
        "type": "interface",
        "uni": false
      },
      "id": "cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 1994
        }
      }
    },
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:02:3",
        "link": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30",
        "lldp": true,
        "mac": "4e:9a:6a:d1:6d:d2",
        "metadata": {},
        "name": "s2-eth3",
        "nni": true,
        "port_number": 3,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:02",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:03:2",
        "link": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30",
        "lldp": true,
        "mac": "fe:6a:fb:24:d7:38",
        "metadata": {},
        "name": "s3-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:03",
        "type": "interface",
        "uni": false
      },
      "id": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 352
        }
      }
    },
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:04:2",
        "link": "88cfed769fbb073a8ca849394180f6b229cbd1cbadd8a1970e6c721ffd675153",
        "lldp": true,
        "mac": "42:24:9c:65:52:d3",
        "metadata": {},
        "name": "s4-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:04",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:03:3",
        "link": "88cfed769fbb073a8ca849394180f6b229cbd1cbadd8a1970e6c721ffd675153",
        "lldp": true,
        "mac": "6a:03:56:41:df:b4",
        "metadata": {},
        "name": "s3-eth3",
        "nni": true,
        "port_number": 3,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:03",
        "type": "interface",
        "uni": false
      },
      "id": "88cfed769fbb073a8ca849394180f6b229cbd1cbadd8a1970e6c721ffd675153",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 2072
        }
      }
    },
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:05:2",
        "link": "0b814adbd3b05669482ca479280787e55be5c155780f8a780423fa9e67e3a848",
        "lldp": true,
        "mac": "22:88:4a:96:c7:06",
        "metadata": {},
        "name": "s5-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:05",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:04:3",
        "link": "0b814adbd3b05669482ca479280787e55be5c155780f8a780423fa9e67e3a848",
        "lldp": true,
        "mac": "0e:77:82:7b:c9:67",
        "metadata": {},
        "name": "s4-eth3",
        "nni": true,
        "port_number": 3,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:04",
        "type": "interface",
        "uni": false
      },
      "id": "0b814adbd3b05669482ca479280787e55be5c155780f8a780423fa9e67e3a848",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 3860
        }
      }
    }
  ],
  "dynamic_backup_path": true,
  "enabled": true,
  "end_date": null,
  "id": "9784971f2e944a",
  "name": "my evc 2000",
  "owner": null,
  "primary_links": [],
  "primary_path": [
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:02:2",
        "link": "cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260",
        "lldp": true,
        "mac": "7e:74:84:8c:62:f5",
        "metadata": {},
        "name": "s2-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:02",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:01:2",
        "link": "cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260",
        "lldp": true,
        "mac": "4a:57:a9:ed:97:6e",
        "metadata": {},
        "name": "s1-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:01",
        "type": "interface",
        "uni": false
      },
      "id": "cf0f4071be426b3f745027f5d22bc61f8312ae86293c9b28e7e66015607a9260",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 1994
        }
      }
    },
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:02:3",
        "link": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30",
        "lldp": true,
        "mac": "4e:9a:6a:d1:6d:d2",
        "metadata": {},
        "name": "s2-eth3",
        "nni": true,
        "port_number": 3,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:02",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:03:2",
        "link": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30",
        "lldp": true,
        "mac": "fe:6a:fb:24:d7:38",
        "metadata": {},
        "name": "s3-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:03",
        "type": "interface",
        "uni": false
      },
      "id": "4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 352
        }
      }
    },
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:04:2",
        "link": "88cfed769fbb073a8ca849394180f6b229cbd1cbadd8a1970e6c721ffd675153",
        "lldp": true,
        "mac": "42:24:9c:65:52:d3",
        "metadata": {},
        "name": "s4-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:04",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:03:3",
        "link": "88cfed769fbb073a8ca849394180f6b229cbd1cbadd8a1970e6c721ffd675153",
        "lldp": true,
        "mac": "6a:03:56:41:df:b4",
        "metadata": {},
        "name": "s3-eth3",
        "nni": true,
        "port_number": 3,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:03",
        "type": "interface",
        "uni": false
      },
      "id": "88cfed769fbb073a8ca849394180f6b229cbd1cbadd8a1970e6c721ffd675153",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 2072
        }
      }
    },
    {
      "active": true,
      "enabled": false,
      "endpoint_a": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:05:2",
        "link": "0b814adbd3b05669482ca479280787e55be5c155780f8a780423fa9e67e3a848",
        "lldp": true,
        "mac": "22:88:4a:96:c7:06",
        "metadata": {},
        "name": "s5-eth2",
        "nni": true,
        "port_number": 2,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:05",
        "type": "interface",
        "uni": false
      },
      "endpoint_b": {
        "active": true,
        "enabled": true,
        "id": "00:00:00:00:00:00:00:04:3",
        "link": "0b814adbd3b05669482ca479280787e55be5c155780f8a780423fa9e67e3a848",
        "lldp": true,
        "mac": "0e:77:82:7b:c9:67",
        "metadata": {},
        "name": "s4-eth3",
        "nni": true,
        "port_number": 3,
        "speed": 1250000000,
        "switch": "00:00:00:00:00:00:00:04",
        "type": "interface",
        "uni": false
      },
      "id": "0b814adbd3b05669482ca479280787e55be5c155780f8a780423fa9e67e3a848",
      "metadata": {
        "s_vlan": {
          "tag_type": 1,
          "value": 3860
        }
      }
    }
  ],
  "priority": -1,
  "queue_id": null,
  "request_time": "2021-10-01T19:35:36",
  "start_date": "2021-10-01T19:35:36",
  "uni_a": {
    "interface_id": "00:00:00:00:00:00:00:01:1",
    "tag": {
      "tag_type": 1,
      "value": 2000
    }
  },
  "uni_z": {
    "interface_id": "00:00:00:00:00:00:00:05:1",
    "tag": {
      "tag_type": 1,
      "value": 2000
    }
  }
}
```

Ping from h1 to h5 will not work as well.

Even though this seems to be related with #40 I think this is a special case, because different from issue #40, this is a valid set of links (returned from Kytos-ng topology!).

### PR DESCRIPTION
**Pull request by ajoaoff**
When creating or updating an EVC, check if given `primary_path` or `backup_path` is valid for this EVC.
Fixes #66

### CODE REVIEW

**Review by viniarck — APPROVED** [review]
Looks good to me, added a suggestion for the error message in the `BadRequest` though. 

@ajoaoff, I also took the chance to review how `Link`(s) are being created, the validation seems safe because once the links are first created they aren't created again, but it seems like we don't have a strong guarantee to create `Link` with deterministic `endpoint_a` and `endpoint_b` , right? Since it's going to depend which switch has sent a Packet In / LLDP packet first https://github.com/kytos-ng/of_lldp/blob/master/main.py#L186-L187, I wonder if in the future it would be worth trying to have some deterministic rules when associating which one should be the endpoint_a and endpoint_b like we have when formatting the Link `id` here https://github.com/kytos-ng/kytos/blob/master/kytos/core/link.py#L72-L83. What do you think, is it worth mapping an issue and continue this thread? If we we're to have this, at least in the output of the `topology` link that would be deterministic as well, even though we know that links are bi-directional.

File: models.py [hunk 1]
```diff
@@ -54,6 +54,17 @@ def make_vlans_available(self):
             link.make_tag_available(link.get_metadata('s_vlan'))
             link.remove_metadata('s_vlan')
 
+    def is_valid(self, switch_a, switch_z):
+        """Check if this is a valid path."""
+        previous = switch_a
+        for link in self:
+            if link.endpoint_a.switch != previous:
```
[the remark this answers is not in the document]
Comment (italovalcy) [reply models.py:61]: @ajoaoff yes, you are right my friend. Although the most common usage will be creating EVCs without schedule, mef_eline does provide such a feature. In this case, the `is_valid()` could receive an additional parameter to enable/disable this link validation check (something like `is_scheduled`).

[the remark this answers is not in the document]
Comment (ajoaoff) [reply models.py:61]: I think it's a good idea.

[the remark this answers is not in the document]
Comment (ajoaoff) [reply models.py:61]: The `link` property in the interface is not being set when a link is created:
https://github.com/kytos-ng/kytos/blob/7862ee9c6c26f4746f241f853e3bcd641a36a530/kytos/core/link.py#L20-L31

I will change it in kytos core before making any change here.

[the remark this answers is not in the document]
Comment (italovalcy) [reply models.py:61]: @ajoaoff as we talked earlier the `link` property will be set by topology napp upon discovering a new link (https://github.com/kytos-ng/topology/blob/master/main.py#L665-L666)

**Comment by italovalcy** [comment]
Looks good to me! Very nice work you have done at this PR, @ajoaoff.

**Review by italovalcy — APPROVED** [review]

**Pull request merged**